In [1]:
import pandas as pd
import numpy as np
from urllib.request import urlretrieve 

# Dataset: Loan Applicant Analysis

In [2]:
# insert code to retrieve dataset from url
url_loans = "https://tinyurl.com/2022aprloans"
df_loans = pd.read_csv(url_loans)

df_loans

Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001003    Male     Yes          1      Graduate            No   
1    LP001005    Male     Yes          0      Graduate           Yes   
2    LP001006    Male     Yes          0  Not Graduate            No   
3    LP001008    Male      No          0      Graduate            No   
4    LP001011    Male     Yes          2      Graduate           Yes   
..        ...     ...     ...        ...           ...           ...   
474  LP002978  Female      No          0      Graduate            No   
475  LP002979    Male     Yes         3+      Graduate            No   
476  LP002983    Male     Yes          1      Graduate            No   
477  LP002984    Male     Yes          2      Graduate            No   
478  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Term  \
0               4583               1508      128000        360   
1               3000                  0       66000        360   
2               2583               2358      120000        360   
3               6000                  0      141000        360   
4               5417               4196      267000        360   
..               ...                ...         ...        ...   
474             2900                  0       71000        360   
475             4106                  0       40000        180   
476             8072                240      253000        360   
477             7583                  0      187000        360   
478             4583                  0      133000        360   

     Credit_History Property_Area Loan_Status  
0                 1         Rural           N  
1                 1         Urban           Y  
2                 1         Urban           Y  
3                 1         Urban           Y  
4                 1         Urban           Y  
..              ...           ...         ...  
474               1         Rural           Y  
475               1         Rural           Y  
476               1         Urban           Y  
477               1         Urban           Y  
478               0     Semiurban           N  

[479 rows x 13 columns]

### Q1a)  2 marks
Load the dataset into a pandas DataFrame <b>df_loans</b>, and display the first 5 rows of the DataFrame.

In [3]:
df_loans.head(5)

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001003   Male     Yes          1      Graduate            No   
1  LP001005   Male     Yes          0      Graduate           Yes   
2  LP001006   Male     Yes          0  Not Graduate            No   
3  LP001008   Male      No          0      Graduate            No   
4  LP001011   Male     Yes          2      Graduate           Yes   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Term  Credit_History  \
0             4583               1508      128000        360               1   
1             3000                  0       66000        360               1   
2             2583               2358      120000        360               1   
3             6000                  0      141000        360               1   
4             5417               4196      267000        360               1   

  Property_Area Loan_Status  
0         Rural           N  
1         Urban           Y  
2         Urban           Y  
3         Urban           Y  
4         Urban           Y

### Q1b)  2 marks
Using the filter method, create a DataFrame <b>df_filtered</b> using only the following columns: Gender, Education and ApplicantIncome. 

Display the first 100 rows of this DataFrame.

Reference: # https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.filter.html

In [4]:
df_filtered = df_loans.filter(items=["Gender", "Education", "ApplicantIncome"])

df_filtered.head(100)

Gender     Education  ApplicantIncome
0    Male      Graduate             4583
1    Male      Graduate             3000
2    Male  Not Graduate             2583
3    Male      Graduate             6000
4    Male      Graduate             5417
..    ...           ...              ...
95   Male      Graduate            10513
96   Male      Graduate             2014
97   Male      Graduate             2718
98   Male      Graduate             3459
99   Male      Graduate             4895

[100 rows x 3 columns]

### Q1c)  4 marks
Create another DataFrame <b>df_urban</b> consisting all columns of the original dataset, but comprises of only applicants in with Urban status in their Property_Area attribute (exclude Rural and Semiurban) with ApplicantIncome of at least S\\$10,000. 

Reset the row index and display the last 10 rows of this DataFrame.

In [5]:
df_urban = df_loans.loc[:, ["Property_Area", "ApplicantIncome"]]
df_urban = df_urban.query("Property_Area == 'Urban' and ApplicantIncome >= 10000")
df_urban = df_urban.reset_index()

df_urban.tail(10)

index Property_Area  ApplicantIncome
6     111         Urban            14866
7     154         Urban            11000
8     199         Urban            16250
9     213         Urban            11146
10    340         Urban            10416
11    378         Urban            18333
12    395         Urban            13262
13    416         Urban            18165
14    446         Urban            16666
15    463         Urban            16120

### Q1d)  2 marks
Create another DataFrame <b>df_income_loan</b> with only the following headers: <b>ApplicantIncome, CoapplicantIncome and LoanAmount</b>.
Use only the .iloc method.

In [6]:
df_income_loan = df_loans.iloc[:, 6:9]

df_income_loan

ApplicantIncome  CoapplicantIncome  LoanAmount
0               4583               1508      128000
1               3000                  0       66000
2               2583               2358      120000
3               6000                  0      141000
4               5417               4196      267000
..               ...                ...         ...
474             2900                  0       71000
475             4106                  0       40000
476             8072                240      253000
477             7583                  0      187000
478             4583                  0      133000

[479 rows x 3 columns]

### Q2a)  5 marks
Out of all the Urban Dwellers with ApplicantIncome of at least S\\$10,000, what is the proportion of customers who are Married?

In [7]:
df_allurban = df_loans.loc[:, ["Property_Area", "ApplicantIncome", "Married"]]
df_allurban = df_loans.query("Property_Area == 'Urban' and ApplicantIncome >= 10000")

allurban_count = df_allurban.shape[0]
df_allurban = df_allurban.query("Married == 'Yes'")
urban_married = len(df_allurban.index)

print(f"The proportion of customers who are married is {urban_married}/{allurban_count}")

The proportion of customers who are married is 10/16


### Q2b)  5 marks
Out of all the Income attributes (i.e. ApplicantIncome and CoapplicantIncome), which column did the highest value come from, and what is this value?

In [8]:
highest_ApplicantIncome = df_income_loan[["ApplicantIncome"]].max(axis=0)["ApplicantIncome"]
highest_CoapplicantIncome = df_income_loan[["CoapplicantIncome"]].max(axis=0)["CoapplicantIncome"]

if highest_ApplicantIncome > highest_CoapplicantIncome:
    highest_IncomeAmt = highest_ApplicantIncome
    print(f"The highest value is stored in column ApplicantIncome and the value is {highest_ApplicantIncome}")
else:
    highest_IncomeAmt = highest_CoapplicantIncome
    print(f"The highest value is stored in column CoapplicantIncome and the value is {highest_CoapplicantIncome}")

The highest value is stored in column ApplicantIncome and the values is 81000


### Q2c)  5 marks
Identify the loan applicants with the highest income amount identified in Q2b.

Comment on the key differences between the profiles of the loan applicants.

In [9]:
df_loans.loc[df_loans["ApplicantIncome"] == highest_IncomeAmt]

Loan_ID  Gender Married Dependents Education Self_Employed  \
27   LP001112  Female     Yes          0  Graduate            No   
318  LP002317    Male     Yes         3+  Graduate            No   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Term  \
27             81000               1459      144000        360   
318            81000                  0      360000        360   

     Credit_History Property_Area Loan_Status  
27                1     Semiurban           Y  
318               0         Rural           N

From the data frame generated, we can see that the applicant with Loan_ID LP001112 is female while the other applicant with Load_ID LP002317 is male. The female applicant is married with no dependents compared to the male
applicant who has more than 3 dependents. Though both applicants have the same income, the female applicant has an additional co-applicant income of $1459. The female applicant also has a smaller loan amount compared to
the male applicant though they have the same loan term, which may be because the female applicant has one credit history already. Another difference is that the female applicant is taking a loan for a semiurban house, 
compared to the male applicant who is taking a loan for his property in th rural area

### Q2d)  10 marks

The column "Loan_Term" measures the duration of the loan applied for, in months.

Applicants asking for loans with loan terms less than 120 months are given a rating of "Short" under loan tenure.
Those with loan terms at least 120 months, but less than 300 months will be rated as "Medium". Applicants asking for loans with loan terms of 300 or more months, will be rated as "Long".

Write a function that takes an applicant's numerical value of months of Loan_Term as input parameter and returns the respective customer's rating. Create a new attribute "Loan_Tenure" for every applicant in df_loans.

Display df_loans with only the "Loan_Term" and "Loan_Tenure" attributes.

In [10]:
# df_loan_termAndTenure = df_loans.loc[:, ["Loan_Term"]]

# #write as function
# df_loan_termAndTenure["Loan_Tenure"] = "Short"
# df_loan_termAndTenure.loc[df_loan_termAndTenure["Loan_Term"] < 300, "Loan_Tenure"] = "Medium"
# df_loan_termAndTenure.loc[df_loan_termAndTenure["Loan_Term"] >= 300, "Loan_Tenure"] = "Long"

# df_loan_termAndTenure
df_loans["Loan_Tenure"] = " "

def check_loan_term(loan_term, index):
    if loan_term < 120:
        df_loans.at[index, "Loan_Tenure"] = "Short"
    elif loan_term < 300:
        df_loans.at[index, "Loan_Tenure"] = "Medium"
    else:
        df_loans.at[index, "Loan_Tenure"] = "Long"

for index, row in df_loans.iterrows():
    current_index = index
    loan_term = row.Loan_Term
    check_loan_term(loan_term, current_index)

df_loans.loc[:, ["Loan_Term", "Loan_Tenure"]]

Loan_Term Loan_Tenure
0          360        Long
1          360        Long
2          360        Long
3          360        Long
4          360        Long
..         ...         ...
474        360        Long
475        180      Medium
476        360        Long
477        360        Long
478        360        Long

[479 rows x 2 columns]

### Q2e)  5 marks
Identify the top 2 applicants with generated the Total Income to Loan Ratio, being Highest Combined Income, (both ApplicantIncome and CoapplicantIncome) divided by Loan Amount.

State one possible issue with the dataset.

In [11]:
#create column for total income to loan ratio (applicant and co-applicant income divided by loan amount)
#find top 2 from the ratio
df_loans["Income_Loan_Ratio"] = 0.0

for index, row in df_loans.iterrows():
    combined_income = row.ApplicantIncome + row.CoapplicantIncome
    loan_amount = row.LoanAmount
    income_to_loan = combined_income / loan_amount
    df_loans.at[index, "Income_Loan_Ratio"] = income_to_loan

df_loans.nlargest(n=2, columns=['Income_Loan_Ratio'])

Loan_ID Gender Married Dependents Education Self_Employed  \
389  LP002604   Male      No          0  Graduate            No   
453  LP002893   Male      No          0  Graduate            No   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Term  \
389             1836              33837       24000        360   
453             1836              33837       24000        360   

     Credit_History Property_Area Loan_Status Loan_Tenure  Income_Loan_Ratio  
389               1         Urban           N        Long           1.486375  
453               1         Urban           N        Long           1.486375

One possible issue with the dataset is that the applicant income is much lower than that of the co-applicant. Additionally, the co-applicant's income is more than sufficient to cover the loan amount, which is evident in 
the income-to-loan ratio which is greater than 1, meaning that the total income from the applicant and co-applicant is sufficient to cover the applied loan amount

### Q2f) 10 marks
The column "LoanAmount" details the amount of loan requested for, in Singapore Dollars, S$. 

Targeted marketing will be conducted for the following groups of customers:

1. Loan Amount: >= S\\$50,000 and  < S\\$250,000
2. Loan Amount: >= S\\$250,000 and  < S\\$450,000

Perform basic statistical analysis and comment on the key differences between these two groups of customers.

In [29]:
#low_loan stores the resultant DataFrame from the provided query expression, which finds the loan applications for loan amounts ranging from $50 000 to less than $250 000
low_loan = df_loans.query("LoanAmount >= 50000 and LoanAmount < 250000")

#The .describe() method generates descriptive statistics and calculates statistical data from the DataFrame stored in low_loan
low_loan.describe()

ApplicantIncome  CoapplicantIncome     LoanAmount   Loan_Term  \
count       418.000000         418.000000     418.000000  418.000000   
mean       4862.763158        1463.696172  130332.535885  343.406699   
std        5191.442345        2033.379585   41952.376149   64.562517   
min         150.000000           0.000000   50000.000000   36.000000   
25%        2807.500000           0.000000  102000.000000  360.000000   
50%        3750.000000        1270.000000  125000.000000  360.000000   
75%        5372.250000        2245.500000  157000.000000  360.000000   
max       81000.000000       20000.000000  246000.000000  480.000000   

       Credit_History  Income_Loan_Ratio  
count      418.000000         418.000000  
mean         0.849282           0.049956  
std          0.358202           0.040557  
min          0.000000           0.014444  
25%          1.000000           0.035676  
50%          1.000000           0.041667  
75%          1.000000           0.050824  
max          1.000000           0.572632

In [28]:
#high_loan stores the resultant DataFrame from the provided query expression, which finds the loan applications for loan amounts ranging from $250 000 to less than $450 000
high_loan = df_loans.query("LoanAmount >= 250000 and LoanAmount < 450000")

#The .describe() method generates descriptive statistics and calculates statistical data from the DataFrame stored in low_loan
high_loan.describe()

ApplicantIncome  CoapplicantIncome     LoanAmount   Loan_Term  \
count        33.000000          33.000000      33.000000   33.000000   
mean      12190.757576        2485.787879  299575.757576  352.727273   
std       13667.951188        3047.242040   47136.126208   32.717802   
min        2661.000000           0.000000  250000.000000  180.000000   
25%        5955.000000           0.000000  260000.000000  360.000000   
50%        8799.000000         240.000000  286000.000000  360.000000   
75%       12841.000000        4232.000000  315000.000000  360.000000   
max       81000.000000       10968.000000  436000.000000  360.000000   

       Credit_History  Income_Loan_Ratio  
count       33.000000          33.000000  
mean         0.878788           0.047964  
std          0.331434           0.037057  
min          0.000000           0.012090  
25%          1.000000           0.034105  
50%          1.000000           0.038996  
75%          1.000000           0.045253  
max          1.000000           0.225000

From the above tables displaying the statistical analysis of DataFrames low_loan and high_loan, we can deduce that there is a larger number of people taking loans which are less than $250 000. Many of those taking smaller loans have lower mean applicant ($4862.76 | $12,190.76) and co-applicant incomes ($1463.70 | $2485.79), and similarly have a lower mean loan amount ($130,332.54 | $299,575.76) as compared to those who take larger loans. Those applying for larger loans tend to have medium to long loan terms, while those who applied for smaller loans tend to have loan terms from as low as 36 months, compared to those who applied for larger loans, having loan terms ranging from 180 months to a maximum of 360 months. However, the overall mean income-to-loan ratio for applicants applying for either smaller or larger loans are almost the same (0.0499 | 0.0480)

## End of Notebook